### ISSUE TITLE
Improve support for staging in F# quotations

### ISSUE THREAD

**Issue by eiriktsarpalis** [fsharp/fslang-suggestions#584]
# Improve support for staging in F# quotations
Staging is a powerful technique for used for type-safe runtime meta-programming. For an introduction to the topic see [this article on MetaOCaml](http://okmij.org/ftp/ML/MetaOCaml.html). Staging can be [somewhat achieved in F# using code quotations](http://fssnip.net/tags/staging). There are however certain important missing pieces.

For example, consider this simple staged power function implementation:
```fsharp
open FSharp.Quotations
open Swensen.Unquote

let rec spower (n : int) : Expr<int> -> Expr<int> =
    if n = 0 then fun _ -> <@ 1 @>
    elif n = 1 then fun t -> <@ %t @>
    else fun x -> <@ %x * (% spower (n-1) x) @>

let eval (stagedComp : Expr<'T> -> Expr<'S>) : Expr<'T -> 'S> =
    <@ fun (t:'T) -> (% stagedComp <@ t @>) @>

spower 10 |> eval |> decompile
// val it : string = "fun t -> t * t * t * t * t * t * t * t * t * t"
```
The problem however with the above is that the `eval` function does not compile, and that is because the compiler won't let us quote the variable `t` when passing to the staged computation. Instead, we are forced to resort to unsafe techniques to achieve the same result:
```fsharp
let eval (stagedComp : Expr<'T> -> Expr<'S>) =
    let var = Var("t", typeof<'T>)
    let body = stagedComp (Expr.Var var |> Expr.Cast)
    Expr.Lambda(var, body) |> Expr.Cast<'T -> 'S>
```
It seems to me that the F# compiler is overly conservative here (since this kind of nested quoting is perfectly possible using MetaOCaml). Would you consider addressing this in future releases of the language?

## Affidavit (please submit!)

Please tick this by placing a cross in the box:
* [x] This is not a question (e.g. like one you might ask on [stackoverflow](http://stackoverflow.com)) and I have searched stackoverflow for discussions of this issue
* [x] I have [searched both open and closed suggestions on this site](http://github.com/fsharp/fslang-suggestions/issues) and believe this is not a duplicate
* [x] This is not something which has obviously "already been decided" in previous versions of F#.  If you're questioning a fundamental design decision that has obviously already been taken (e.g. "Make F# untyped") then please don't submit it.

Please tick all that apply:
* [ ] This is not a breaking change to the F# language design
* [ ] I or my company would be willing to help implement and/or test this

**Comment by dsyme** [comment]
@eiriktsarpalis Yes, I think it's reasonable. It may depend on the complexity of the implementation though.  Given the current implementation of splicing, I'm not sure how we would build ``<@ t @>`` to pass to ``stagedComp `` while making the splice argument - the object representing ``t`` only gets created after during deserialization, which happens after preparing the splice argument...

**Comment by eiriktsarpalis** [comment]
@dsyme perhaps quoted bindings that escape their scope could be passed to
the unpickling function parametrically, just like System.Type instances
currently.

**Comment by dsyme** [comment]
@eiriktsarpalis Yes that might work.

**Comment by eiriktsarpalis** [comment]
@dsyme Would you consider merging a PR that makes this improvement?

**Comment by dsyme** [comment]
> @dsyme Would you consider merging a PR that makes this improvement?

Yes, subject to usual caveats (testing, RFC, complexity of implementation etc.)

**Comment by dsyme** [comment]
@eiriktsarpalis Any progress on this?

**Comment by gsomix** [comment]
@dsyme Is it approved in principle?

**Comment by dsyme** [comment]
@gsomix Yes, it's improved in principle. But I don't think anyone is working on this one.

**Comment by eiriktsarpalis** [comment]
I'm sorry, but I haven't time to work on this at all. In hindsight the lack on efficient quotation compiler is the bigger blocker when it comes to staged programming.

**Comment by 7sharp9** [comment]
Does this need an RFC?

**Comment by dsyme** [comment]
@7sharp9 I think it needs a champion who would be prepared to drive it all the way through to implementation

**Comment by cartermp** [comment]
Re-labelling here, unless `approved-in-principle` is redefined to mean, "has a champion willing to drive the suggestion all the way through to implementation". We have many issues tagged here that have no RFC nor champion, and given this was previously labeled as such I'm re-applying the label

**Comment by 7sharp9** [comment]
Maybe a rough RFC should be approved in principle first?

Would type splicing make this more viable or easier to complete?
